dbms/src/Flash/Mpp/ReceiverChannelBase.cpp

```diff
@@ -19,47 +19,66 @@
-    RUNTIME_ASSERT(packet.chunks_size() == packet.stream_ids_size(), log, "packet's chunk size shoule be equal to it's size of streams");
-
-    for (int i = 0; i < packet.stream_ids_size(); ++i)
-    {
-        UInt64 stream_id = packet.stream_ids(i) % channel_size;
-        chunks[stream_id].push_back(&packet.chunks(i));
-    }
-
-    return true;
-}
-
-const mpp::Error * ReceiverChannelBase::getErrorPtr(const mpp::MPPDataPacket & packet)
+const mpp::Error * getErrorPtr(const mpp::MPPDataPacket & packet)
 {
     if (unlikely(packet.has_error()))
         return &packet.error();
     return nullptr;
 }
 
-const String * ReceiverChannelBase::getRespPtr(const mpp::MPPDataPacket & packet)
+const String * getRespPtr(const mpp::MPPDataPacket & packet)
 {
     if (unlikely(!packet.data().empty()))
         return &packet.data();
     return nullptr;
 }
+} // namespace
+
+ReceivedMessagePtr toReceivedMessage(
+    const TrackedMppDataPacketPtr & tracked_packet,
+    size_t source_index,
+    const String & req_info,
+    bool for_fine_grained_shuffle,
+    size_t fine_grained_consumer_size)
+{
+    const auto & packet = tracked_packet->packet;
+    const mpp::Error * error_ptr = getErrorPtr(packet);
+    const String * resp_ptr = getRespPtr(packet);
+    std::vector<const String *> chunks(packet.chunks_size());
+    for (int i = 0; i < packet.chunks_size(); ++i)
+        chunks[i] = &packet.chunks(i);
+    auto ret = std::make_shared<ReceivedMessage>(
+        source_index,
+        req_info,
+        tracked_packet,
+        error_ptr,
+        resp_ptr,
+        std::move(chunks));
+    if (for_fine_grained_shuffle)
+    {
+        assert(fine_grained_consumer_size > 0);
+        ret->remaining_consumers = std::make_shared<std::atomic<size_t>>(fine_grained_consumer_size);
+        ret->fine_grained_chunks.resize(fine_grained_consumer_size);
+        if (packet.chunks_size() > 0)
+        {
+            RUNTIME_CHECK_MSG(!packet.stream_ids().empty(), "MPPDataPacket.stream_ids empty, it means ExchangeSender is old version of binary "
+                                                            "(source_index: {}) while fine grained shuffle of ExchangeReceiver is enabled. "
+                                                            "Cannot handle this.",
+                              source_index);
+
+            // packet.stream_ids[i] is corresponding to packet.chunks[i],
+            // indicating which stream_id this chunk belongs to.
+            RUNTIME_CHECK_MSG(packet.chunks_size() == packet.stream_ids_size(), "Packet's chunk size({}) not equal to its size of streams({})", packet.chunks_size(), packet.stream_ids_size());
+
+            for (int i = 0; i < packet.stream_ids_size(); ++i)
+            {
+                UInt64 stream_id = packet.stream_ids(i) % fine_grained_consumer_size;
+                ret->fine_grained_chunks[stream_id].push_back(&packet.chunks(i));
+            }
+        }
+    }
```

Comment:
Looks like it could be wrapped in a ReceivedMessage

Reply:
Done